### ISSUE TITLE
Memory safety story?

### ISSUE THREAD

**Issue by asb** [scrossuk/locic#9]
Hi Stephen, I was wondering if you had any thoughts or plans for a memory safety story for Loci? The planned support for garbage collection may obviously help in some cases. I was wondering if you were following the work on the C++ core guidelines, or perhaps previous efforts such as [Ironclad C++](http://acg.cis.upenn.edu/papers/ironclad-oopsla.pdf). I understand that there's a risk of violating some of Loci's core design principles, but for many use cases a greater degree of confidence about memory safety can improve programmer productivity.

**Comment by scrossuk** [comment]
Hi Alex.

> The planned support for garbage collection may obviously help in some cases.

Actually I'm not much of a fan of garbage collection. Like you, I'm more interested in the approach taken by Rust and C++ core guidelines, because it's less costly, more precise and supports deterministic destruction. My only interest in garbage collection is for some cases where it can be cheaper than reference counting, and where destruction isn't required.

>  was wondering if you were following the work on the C++ core guidelines, or perhaps previous efforts such as Ironclad C++.

Yes, absolutely. I've been thinking about what the best solution is, and leaning more towards the way Rust handles this (i.e. at any time only either one mutable reference or any number of const references), though I'm also watching the C++ side.

> I understand that there's a risk of violating some of Loci's core design principles, but for many use cases a greater degree of confidence about memory safety can improve programmer productivity.

Yes, I agree, and to me this is a core language issue. It's just a matter of having enough time to work on this :-).

**Comment by asb** [comment]
> Yes, absolutely. I've been thinking about what the best solution is, and leaning more towards the way Rust handles this (i.e. at any time only either one mutable reference or any number of const references), though I'm also watching the C++ side.

I agree, that a solution that either doesn't involve GC or minimises the need for it is much more in the spirit of Loci being a better C++. My only concern about the Rust model is the mental overhead. Ironclad is simpler in some respects, but less flexible (no explicit lifetime tracking so pointers are either heap/global or heap/global/local).

Pony is worth a look if you haven't seen it, though if anything manages to have higher cognitive overhead than Rust. Although coming from a very different direction (Scala), [this paper](https://www.cs.purdue.edu/homes/rompf/papers/osvald-oopsla16.pdf) made a nice argument for the benefits of second-class values, exploring the uses beyond memory management (e.g. capabilities).
